server/s3_motr_writer.cc

```diff
@@ -473,6 +474,30 @@ void S3MotrWiter::write_content() {
   ctx->cbs[0].oop_stable = s3_motr_op_stable;
   ctx->cbs[0].oop_failed = s3_motr_op_failed;
   set_up_motr_data_buffers(rw_ctx, data_items, motr_buf_count);
+
+  if (S3Option::get_instance()->get_s3_st_md5_error_inject_enabled()) {
+    struct m0_bufvec *bv = rw_ctx->data;
+    if (rw_ctx->ext->iv_index[0] == 0) {
+      char first_byte = *(char *)bv->ov_buf[0];
+      s3_log(S3_LOG_DEBUG, "", "%s first_byte=%d", __func__, first_byte);
+      switch (first_byte) {
+        case 'z':  // zero
+	  corrupt_fill_zero = true;
+          break;
+        case 'f':  // first
+          // corrupt the first byte
+          *(char *)bv->ov_buf[0] = ~first_byte;
+          break;
+        case 'k':  // OK
+          break;
+      }
+    }
+    if (corrupt_fill_zero) {
+      for (uint32_t i = 0; i < bv->ov_vec.v_nr; ++i)
+        memset(bv->ov_buf[i], 0, bv->ov_vec.v_count[i]);
```

Comment:
Is this the function that injects errors?  Why are there two different ways to inject failures?  Seems unnecessarily complicated.  Just typecast the buffer to a char array and increment the first char.  Simple, easy, foolproof. No?

Reply:
2 kinds of corruption were added: change of the first char and zeroing the buffers to ensure that a) checksums are calculated for the entire object and b) change of each byte would be detected.

Reply:
> Just typecast the buffer to a char array and increment the first char. 

This is also done a few lines earlier. Not increment, because increment of a char (that might be signed) is not something I'd like to have because of 6.3.1.3/3 in C99 ([n1256.pdf](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf)):

> Otherwise, the new type is signed and the value cannot be represented in it; either the result is implementation-defined or an implementation-defined signal is raised.

I just checked the code that is currently being reviewed and it uses `~` which may also have issues. I changed it to `!`, which results in either 0 or 1, which should be fine for any type. Thanks for pointing out, it helped to find another issue.